### ISSUE TITLE
Support connection upgrade using `rack.hijack` partial hijack.

### ISSUE THREAD

**Issue by ioquatix** [puma/puma#2914]
It would be nice if Puma can support WebSockets, even if it was limited to one WebSocket per thread. Allocation 100 or 1000 threads per instance isn't a bad model for some use cases.

I've been playing around with this, but can't get it to work.

I have a sample server:

```ruby
require 'async/websocket/adapters/rack'

app = lambda do |env|
  Async::WebSocket::Adapters::Rack.open(env) do |connection|
    while message = connection.read
      connection.write message
    end
  end or [404, {}, []]
end

run app
```

This returns a response which looks like this:

```
[101,
 {"sec-websocket-extensions"=>"permessage-deflate;client_max_window_bits=15",
  "sec-websocket-accept"=>"1q4ves5M1K+fV+tvlOTSdG9WzZQ=",
  "upgrade"=>"websocket",
  "rack.hijack"=>
   #<Async::HTTP::Body::Hijack #<Proc:0x000000010849da78 /Users/samuel/Projects/socketry/async-websocket/lib/async/websocket/adapters/http.rb:53>>},
 []]
```

I have a proposal to [support a better upgrade mechanism](https://github.com/rack/rack/issues/1946), but in theory, this should be sufficient. However, I get the following error:

```
2022-08-22 21:47:19 +1200 HTTP parse error, malformed request: #<Puma::HttpParserError: Invalid HTTP format, parsing fails. Are you trying to open an SSL connection to a non-SSL Puma?>
```

I'm guessing something is wrong at the protocol level. `rack.hijack` never seems to be called. I wonder if Puma is trying to buffer all the input? `rack.hijack` probably expects streaming input.

**Comment by ioquatix** [comment]
I made a minimal (non-web socket) server just to see what I could trigger:

```ruby
app = lambda do |env|
	p env
	callback = proc do |stream|
		binding.irb
	end
	
	[101, {'rack.hijack' => callback, 'upgrade' => 'websocket'}, []]
end
```

I noticed that even for an incoming connection with `upgrade: web socket`, `env` contains `"rack.input"=>#<Puma::NullIO:0x0000000104bf3830>` which seems... incorrect? Surely we can read from the input stream here?

**Comment by MSP-Greg** [comment]
@ioquatix

Re the above, were you using a TLSv1.3 connection?  Could you try it forced to TLSv1.2?

**Comment by ioquatix** [comment]
@MSP-Greg I believe I was just using plain HTTP, no TLS. But I can double check.

**Comment by MSP-Greg** [comment]
Could something have used port 443?  I'll check the code, but that error only happens when an http connection is made to a server setup to use only SSL clients, and also with TLSv1.3 as the negotiated protocol?

**Comment by ioquatix** [comment]
https://github.com/ioquatix/puma-switching-protocols

I was trying to get this to work. I can't seem to get the `rack.hijack` callback invoked. Did I do something wrong?

**Comment by dentarg** [comment]
> https://github.com/ioquatix/puma-switching-protocols

If I change the status code from `101` to `200` the callback is invoked (I don't know why) Saw it here: https://github.com/puma/puma/blob/f59887602c66ad1f257ad57ab3aa27214b133c11/test/test_puma_server.rb#L580-L595

**Comment by ioquatix** [comment]
Oh, great, let me try it.

**Comment by ioquatix** [comment]
I've created https://github.com/rack/rack/pull/1954 as a proposal for how we can solve protocol upgrades with a shared semantic between HTTP/1 and HTTP/2. While we still need to do protocol specific things, actually handling the streaming is a bit easier with shared understand of "this request wants to use the following protocol, and this response will use that protocol".

**Comment by MSP-Greg** [comment]
A long time ago, I knew the differences between the draft WebSocket specs/recs.  I'm interested in this, but I need a big refresh on the spec.

Does WebSocket exist with HTTP/3?  Given that it is a different protocol (UDP), not sure...

**Comment by wtn** [comment]
In my testing, `Async::WebSocket::Adapters::Rails#open` works in Puma 6.1.0 and later. From the commit logs, it appears the fix came in 6da32ca82e6f3089afd61b92dbe211fa5e99f220.

Can this issue be closed?

**Comment by ioquatix** [comment]
Yes, thanks for checking!
